### PR TITLE
Improve `assert_valid` for capture

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**

Changes extracted from https://github.com/PennyLaneAI/pennylane/pull/10019

**Description of the Change:**

- General code re-organization and cleanups
- Update wire-mapping tests to use integers instead of strings for wire labels
- Removes `skip_capture`. Capture tests are now run only if capture is enabled.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-128641]

### CODE REVIEW

File: pennylane/ops/functions/assert_valid.py [hunk 1]
```diff
@@ -223,6 +232,34 @@ def _assert_counts_match(counts_0, counts_1):
...
+    else:
+        # TODO: tracing Operator2 hybrid args is not supported due to [sc-127789], move
+        # op.hybrid_args to capture_kwargs when the issue with wires is fixed.
+        decomposition = partial(rule, **op.static_args, **op.compilable_args, **op.hybrid_args)
+        capture_args = ()
+        capture_kwargs = {**op.dynamic_args, **op.wire_args}
+
+    plxpr = qp.capture.make_plxpr(decomposition, autograph=False)(*capture_args, **capture_kwargs)
+    flat_capture_args = jax.tree.leaves((capture_args, capture_kwargs))
+    return qp.tape.plxpr_to_tape(plxpr.jaxpr, plxpr.consts, *flat_capture_args)
```
Comment (mudit2812) [inline pennylane/ops/functions/assert_valid.py:280]: I think there is a really hacky way to get around the hybrid arguments issue. I will try it out tomorrow and let you know if it works

File: pennylane/ops/functions/assert_valid.py [hunk 2]
```diff
@@ -489,18 +498,16 @@ def _check_pytree(op):
...
         ) from e
-    try:
-        import jax
-    except ImportError:
-        return
-    leaves, struct = jax.tree_util.tree_flatten(op)
-    unflattened_op = jax.tree_util.tree_unflatten(struct, leaves)
+
+    leaves, struct = qp.pytrees.flatten(op)
+    unflattened_op = qp.pytrees.unflatten(leaves, struct)
```
Comment (mudit2812) [inline pennylane/ops/functions/assert_valid.py]: I think that if jax is available then we should also try jax flattening/unflattening

**Review by mudit2812 — COMMENTED** [review]
Just the one comment. Otherwise looks great

File: pennylane/ops/functions/assert_valid.py (hunk 2 above)
Comment (astralcai) [reply pennylane/ops/functions/assert_valid.py]: [9083026](https://github.com/PennyLaneAI/pennylane/pull/10058/commits/9083026cb68aaae9054a981fd12aee863c58d811)

**Review by JerryChen97 — APPROVED** [review]

**Comment by mudit2812** [comment]
@astralcai this is what I had in mind. It's ugly, but flattening will get rid of wires being unflattened with `ArgInfo` at tracing boundaries. I prefer we handle hybrid arguments correctly now rather than waiting for `Wires` to get fixed because I really don't know when we'll be able to get to it.
```python
def _decomposition(*args, **kwargs):
    # This is the decomp rule with static/compilable args
    # partialled away
    return


# Dummy hybrid args
hybrid_args = {"a": [1, 2, 3], "b": qp.RX(1.5, 0)}

capture_args = ()
capture_kwargs = {"c": 1.5, "d": 2.5}

hybrid_trees = {}
for name, val in hybrid_args.items():
    leaves, tree = qp.pytrees.flatten(val)
    capture_kwargs[name] = leaves
    hybrid_trees[name] = tree


def decomposition(*args, **kwargs):
    for h in hybrid_args:
        kwargs[h] = qp.pytrees.unflatten(kwargs[h], hybrid_trees[h])
    _decomposition(*args, **kwargs)


jaxpr = jax.make_jaxpr(decomposition)(*capture_args, **capture_kwargs)
```

**Comment by astralcai** [comment]
> @astralcai this is what I had in mind. It's ugly, but flattening will get rid of wires being unflattened with `ArgInfo` at tracing boundaries. I prefer we handle hybrid arguments correctly now rather than waiting for `Wires` to get fixed because I really don't know when we'll be able to get to it.
> 
> ```python
> def _decomposition(*args, **kwargs):
>     # This is the decomp rule with static/compilable args
>     # partialled away
>     return
> 
> 
> # Dummy hybrid args
> hybrid_args = {"a": [1, 2, 3], "b": qp.RX(1.5, 0)}
> 
> capture_args = ()
> capture_kwargs = {"c": 1.5, "d": 2.5}
> 
> hybrid_trees = {}
> for name, val in hybrid_args.items():
>     leaves, tree = qp.pytrees.flatten(val)
>     capture_kwargs[name] = leaves
>     hybrid_trees[name] = tree
> 
> 
> def decomposition(*args, **kwargs):
>     for h in hybrid_args:
>         kwargs[h] = qp.pytrees.unflatten(kwargs[h], hybrid_trees[h])
>     _decomposition(*args, **kwargs)
> 
> 
> jaxpr = jax.make_jaxpr(decomposition)(*capture_args, **capture_kwargs)
> ```

[70c5d9b](https://github.com/PennyLaneAI/pennylane/pull/10058/commits/70c5d9b9946babad956ba9494b3dc06d347a019a)

File: pennylane/ops/functions/assert_valid.py [hunk 3]
```diff
@@ -223,6 +232,54 @@ def _assert_counts_match(counts_0, counts_1):
...
+    if isinstance(op, Operator1):
+        decomposition = partial(rule, **op.hyperparameters)
+        capture_args = op.data
+        capture_kwargs = {"wires": op.wires}
+    else:
+        decomposition = partial(rule, **op.static_args, **op.compilable_args)
+        capture_args = ()
+        capture_kwargs = {**op.dynamic_args, **op.wire_args}  # hybrid args will be added below
+        # TODO: tracing Operator2 hybrid args is not supported out of the box due to [sc-127789].
+        # For now we manually flatten the hybrid args and assembles them within the wrapper rule,
```
Comment (mudit2812) [inline pennylane/ops/functions/assert_valid.py]:
```suggestion
        # For now we manually flatten the hybrid args and reassemble them within the wrapper rule,
```

File: tests/ops/functions/test_assert_valid.py [hunk 4]
```diff
@@ -25,10 +26,9 @@
 
 import pennylane as qp
 from pennylane.core import Operator2
-from pennylane.core.operator import Operator
+from pennylane.core.operator import Operator, Operator1
```
Comment (mudit2812) [inline tests/ops/functions/test_assert_valid.py]: If you're just using `Operator1` to create a class for testing, you can just use `Operator`
```suggestion
from pennylane.core.operator import Operator
```

File: tests/ops/functions/test_assert_valid.py [hunk 5]
```diff
@@ -815,6 +799,16 @@ def __init__(self, phi, wires, ops):
             )
 
 
+@pytest.mark.capture
+def test_op1_assert_valid_capture():
+    """Tests calling assert_valid on Operator1 with capture enabled."""
+
+    class CustomOp(Operator1):
```
Comment (mudit2812) [inline tests/ops/functions/test_assert_valid.py]:
```suggestion
    class CustomOp(Operator):
```

File: tests/ops/functions/test_assert_valid.py (hunk 4 above)
Comment (mudit2812) [reply tests/ops/functions/test_assert_valid.py]: Non-blocking though

**Review by mudit2812 — APPROVED** [review]
